### PR TITLE
AMDGPU: Pre-commit test to verify mode change in fp constrained operations

### PR DESCRIPTION
**Pull request by abhigargrepo**
This test will check the  mode register in case of constrained floating point operations.

### CODE REVIEW

**Review by arsenm — COMMENTED** [review]
This can be an IR test. Commit message should also be more specific

File: llvm/test/CodeGen/AMDGPU/mode-register-fpconstrain.ll [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+; GCN-NEXT:    v_add_f64 v[0:1], v[0:1], v[2:3]
+; GCN-NEXT:    s_nop 0
+; GCN-NEXT:    s_setreg_imm32_b32 hwreg(HW_REG_MODE, 2, 2), 0
+; GCN-NEXT:    s_setpc_b64 s[30:31]
+  tail call void @llvm.amdgcn.s.setreg(i32 2177, i32 1)
+  %3 = tail call double @llvm.experimental.constrained.fadd.f64(double %0, double %1, metadata !"round.dynamic", metadata !"fpexcept.strict") #0
+  tail call void @llvm.amdgcn.s.setreg(i32 2177, i32 0)
+  ret double %3
+}
+
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/mode-register-fpconstrain.ll:23]: Can you also add a test function using the standard llvm.set.fpenv intrinsics (and maybe set.rounding but then you depend on the PR to support that)

File: llvm/test/CodeGen/AMDGPU/mode-register-fpconstrain.ll [hunk 2]
```diff
@@ -0,0 +1,28 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 4
+; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 -verify-machineinstrs %s -o - | FileCheck -check-prefix=GCN %s
+
+; The si-mode-register pass is changing the default mode for FP constrained  operations.
+; It must ignore for strictfp functions.
+
+define hidden fastcc double @ignoreStrictfp(double noundef %0, double noundef %1) unnamed_addr #0 {
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/mode-register-fpconstrain.ll:7]:
```suggestion
define double @ignoreStrictfp(double noundef %0, double noundef %1) #0 {
```

File: llvm/test/CodeGen/AMDGPU/mode-register-fpconstrain.ll [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+; GCN-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
+; GCN-NEXT:    s_setreg_imm32_b32 hwreg(HW_REG_MODE, 2, 2), 1
+; GCN-NEXT:    s_nop 1
+; GCN-NEXT:     s_setreg_imm32_b32 hwreg(HW_REG_MODE, 2, 1), 0
+; GCN-NEXT:    v_add_f64 v[0:1], v[0:1], v[2:3]
+; GCN-NEXT:    s_nop 0
+; GCN-NEXT:    s_setreg_imm32_b32 hwreg(HW_REG_MODE, 2, 2), 0
+; GCN-NEXT:    s_setpc_b64 s[30:31]
+  tail call void @llvm.amdgcn.s.setreg(i32 2177, i32 1)
+  %3 = tail call double @llvm.experimental.constrained.fadd.f64(double %0, double %1, metadata !"round.dynamic", metadata !"fpexcept.strict") #0
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/mode-register-fpconstrain.ll:19]: Use named values in tests

**Review by arsenm — COMMENTED** [review]
Patch title should be made more specific for what this is a test for

File: llvm/test/CodeGen/AMDGPU/mode-register-fpconstrain.ll (hunk 1 above)
Comment (abhigargrepo) [reply llvm/test/CodeGen/AMDGPU/mode-register-fpconstrain.ll:36]: Added a test function to check for llvm.set.fpenv intrinsics

File: llvm/test/CodeGen/AMDGPU/mode-register-fpconstrain.ll [hunk 4]
```diff
@@ -0,0 +1,43 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 4
+; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 -verify-machineinstrs %s -o - | FileCheck -check-prefix=GCN %s
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/mode-register-fpconstrain.ll:2]:
```suggestion
; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 %s -o - | FileCheck -check-prefix=GCN %s
```

File: llvm/test/CodeGen/AMDGPU/mode-register-fpconstrain.ll [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+; GCN:       ; %bb.0: ; %entry
+; GCN-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
+; GCN-NEXT:    s_setreg_imm32_b32 hwreg(HW_REG_MODE, 0, 23), 4
+; GCN-NEXT:    s_setreg_imm32_b32 hwreg(HW_REG_TRAPSTS, 0, 5), 0
+; GCN-NEXT:    s_nop 0
+; GCN-NEXT:    s_setreg_imm32_b32 hwreg(HW_REG_MODE, 2, 1), 0
+; GCN-NEXT:    v_add_f64 v[0:1], v[0:1], v[2:3]
+; GCN-NEXT:    s_setpc_b64 s[30:31]
+entry:
+  call void @llvm.set.fpenv.i64(i64 4)
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/mode-register-fpconstrain.ll:32]: Not the most realistic source value but ok

**Review by arsenm — APPROVED** [review]

**Pull request merged**